### ISSUE TITLE
support OWL/XML format in upload operation (was: errors uploading ice-of-land-origin ontology)

### ISSUE THREAD

**Issue by carueda** [mmisw/mmiorr#349]
Reported by @graybeal .

Original file: from https://ssiii.googlecode.com/svn/trunk/ontology/ice-of-land-origin.owl

I reproduced the errors locally while monitoring the log file:

- Use Upload option in the ORR
- indicate the file and click upload
- the log file shows errors like the following
```
2015-04-20 20:16:28,246 ERROR [com.hp.hpl.jena.rdf.model.impl.RDFDefaultErrorHandler] - file:/Users/Shared/mmiregistry/ontmd/preuploads/ontmd_EC208D05E92089E85F9818D874A10951_3536197894802071235.tmp(line 35 column 50): {E201} Multiple children of property element
2015-04-20 20:16:28,246 ERROR [com.hp.hpl.jena.rdf.model.impl.RDFDefaultErrorHandler] - file:/Users/Shared/mmiregistry/ontmd/preuploads/ontmd_EC208D05E92089E85F9818D874A10951_3536197894802071235.tmp(line 35 column 99): {E202} Expecting XML start or end element(s). String data "http://creativecommons.org/publicdomain/zero/1.0/" not allowed. Maybe there should be an rdf:parseType='Literal' for embedding mixed XML content in RDF. Maybe a striping error.
2015-04-20 20:16:28,247 ERROR [com.hp.hpl.jena.rdf.model.impl.RDFDefaultErrorHandler] - file:/Users/Shared/mmiregistry/ontmd/preuploads/ontmd_EC208D05E92089E85F9818D874A10951_3536197894802071235.tmp(line 39 column 50): {E201} Multiple children of property element
2015-04-20 20:16:28,247 ERROR [com.hp.hpl.jena.rdf.model.impl.RDFDefaultErrorHandler] - file:/Users/Shared/mmiregistry/ontmd/preuploads/ontmd_EC208D05E92089E85F9818D874A10951_3536197894802071235.tmp(line 39 column 53): {E202} Expecting XML start or end element(s). String data "1.0" not allowed. Maybe there should be an rdf:parseType='Literal' for embedding mixed XML content in RDF. Maybe a striping error.
2015-04-20 20:16:28,247 ERROR [com.hp.hpl.jena.rdf.model.impl.RDFDefaultErrorHandler] - file:/Users/Shared/mmiregistry/ontmd/preuploads/ontmd_EC208D05E92089E85F9818D874A10951_3536197894802071235.tmp(line 43 column 14): {E201} Multiple children of property element
2015-04-20 20:16:28,247 ERROR [com.hp.hpl.jena.rdf.model.impl.RDFDefaultErrorHandler] - file:/Users/Shared/mmiregistry/ontmd/preuploads/ontmd_EC208D05E92089E85F9818D874A10951_3536197894802071235.tmp(line 43 column 97): {E202} Expecting XML start or end element(s). String data "http://dataportal.ucar.edu/schemas/gcmd.owl#cryosphere_glaciers_ice_sheets_glaciers" not allowed. Maybe there should be an rdf:parseType='Literal' for embedding mixed XML content in RDF. Maybe a striping error.
...
2015-04-20 20:16:28,278 ERROR [com.hp.hpl.jena.rdf.model.impl.RDFDefaultErrorHandler] - file:/Users/Shared/mmiregistry/ontmd/preuploads/ontmd_EC208D05E92089E85F9818D874A10951_3536197894802071235.tmp(line 440 column 50): {E201} Multiple children of property element
2015-04-20 20:16:28,278 ERROR [com.hp.hpl.jena.rdf.model.impl.RDFDefaultErrorHandler] - file:/Users/Shared/mmiregistry/ontmd/preuploads/ontmd_EC208D05E92089E85F9818D874A10951_3536197894802071235.tmp(line 440 column 62): {E202} Expecting XML start or end element(s). String data "Tabular berg" not allowed. Maybe there should be an rdf:parseType='Literal' for embedding mixed XML content in RDF. Maybe a striping error.

2015-04-20 20:23:27,544 DEBUG [org.mmisw.orrclient.core.OrrClientImpl] - _getOntologyContents(TempOntologyInfo): getting metadata
2015-04-20 20:23:27,544 DEBUG [org.mmisw.orrclient.core.MetadataExtractor] - prepareOntologyMetadata: ontologyUri=http://www.w3.org/2002/07/owl

2015-04-20 20:23:27,546 DEBUG [org.mmisw.orrclient.core.OrrClientImpl] - _getOntologyContents(TempOntologyInfo): getting entities
2015-04-20 20:23:27,551 DEBUG [org.mmisw.orrclient.core.util.ontinfo.OntInfo] - subjects=0 individuals=0 classes=0 properties=0
2015-04-20 20:23:27,551 DEBUG [org.mmisw.orrclient.core.util.ontype.OntTypeUtil] - determineType: ontUri=(ont is null) ontologyUri=http://www.w3.org/2002/07/owl --> other
```

The last few lines above seem to indicate the file does not have a proper URI for the ontology itself defined, but somehow the tool takes `http://www.w3.org/2002/07/owl` as that URI..

- the UI does not give any indication of errors so far
- continue upload sequence (choosing "re-hosted" as hosting mode) and click finish.  Error Alert occurs: 
![image](https://cloud.githubusercontent.com/assets/556505/7245099/7155f68e-e79a-11e4-95b3-6c278f771d17.png)
- and log file shows similar errors as in initial file upload, and also:
```
2015-04-20 20:16:28,280 DEBUG [org.mmisw.ont.JenaUtil2] - Returning Ontology with URI: null
2015-04-20 20:16:28,281 DEBUG [org.mmisw.ont.JenaUtil2] - Returning Ontology with URI: null
2015-04-20 20:16:28,281 INFO  [org.mmisw.orrclient.core.OrrClientImpl] - createOntologyReHosted: original namespace: null
2015-04-20 20:16:28,281 DEBUG [org.mmisw.orrclient.core.OrrClientImpl] - Getting pre-existing properties from Ontology: null
2015-04-20 20:16:28,282 DEBUG [org.mmisw.orrclient.core.ontmodel.OntModelUtil] - New ontology created with namespace http://www.w3.org/2002/07/owl/ base http://www.w3.org/2002/07/owl
2015-04-20 20:16:28,282 DEBUG [org.mmisw.ont.JenaUtil2] - WARNING: more than one Ontology resource in OntModel. Second one found (but ignored): null
2015-04-20 20:16:28,282 DEBUG [org.mmisw.ont.JenaUtil2] - Returning Ontology with URI: http://www.w3.org/2002/07/owl
2015-04-20 20:16:28,283 DEBUG [org.mmisw.orrclient.core.OrrClientImpl] - New ontology created with namespace http://www.w3.org/2002/07/owl/ base http://www.w3.org/2002/07/owl
```

**Comment by carueda** [comment]
Now, I just tried the following:

- Loaded the original ontology using Protege (4.1.0)
- The OWL/XML rendering is what corresponds to the original file
- but I copied the RDF/XML rendering into a new file `ice-of-land-origin.rdf`
- Retried with this new file:
- Clicked upload and no errors on the log with last couple lines reading:
```
2015-04-20 22:57:42,887 DEBUG [org.mmisw.orrclient.core.OrrClientImpl] - _getOntologyContents(TempOntologyInfo): getting entities
2015-04-20 22:57:42,888 DEBUG [org.mmisw.orrclient.core.util.ontinfo.OntInfo] - subjects=32 individuals=0 classes=11 properties=1
2015-04-20 22:57:42,888 DEBUG [org.mmisw.orrclient.core.util.ontype.OntTypeUtil] - determineType: ontUri=(ont is null) ontologyUri=http://purl.org/wmo/seaice/iceOfLandOrigin --> other
```
which seem ok.

- the overall upload sequence (on my local system and in re-hosted mode) completes fine.
![image](https://cloud.githubusercontent.com/assets/556505/7245490/dfde3458-e7a0-11e4-82cd-d273b229b77e.png)


So, @graybeal , while we get to investigate further, perhaps you want to use this workaround, that is, use the RDF/XML rendering as the source for the upload in the ORR..

**Comment by carueda** [comment]
Ah! just found: http://answers.semanticweb.com/questions/15004/using-jena-to-access-an-owlxml-file-made-from-protege

**Comment by carueda** [comment]
Sorry, seems like I also uploaded this test to mmisw ... just unregistered it.

**Comment by graybeal** [comment]
Heh!  OK, so I get the gist -- I wonder if the xmllint suggestion in the semanticweb page above (`xmllint --noent owlfile -o rdfxmlfile`) would be enough to make it work? Might try that.

The Protege rendering trick is pretty neat, and I couldn't figure out why I had no idea what you were talking about. I found [this page](http://protegewiki.stanford.edu/wiki/Pr4_UG_rv_On_RDF_XML_rend) which was instructive, because my view of the Active Ontology view doesn't have those two sub-views. I found the article [here](http://protege-project.136.n4.nabble.com/Re-File-Owl-protege-discussion-Digest-Vol-76-Issue-26-td4656641.html) which says 

```
Not sure if this is what you want but the Protege 4 tab for the "Active 
Ontology" can be configured to show many reifications of the currently 
loaded ontology. You can configure the tab to show; 
- RDF/XML rendering 
- OWL functional syntax rendering 
- OWL/XML rendering 
<snip>
Go to <Window><views><ontology views> and select the reification you 
want to include and place it in the window for the tab. I put them on 
the "Active Ontology" tab but you could make your own tab with the 
various reifications. 
```

So I can go there but the act of 'put them' escapes me -- all I can do is select the appropriate option, but nothing happens as a result.

Probably a Protege mail list question, I'm not seeing it on the web.

**Comment by graybeal** [comment]
xmllint did not do the trick. (too bad, so sad.)

**Comment by carueda** [comment]
Just marked this as a possible *enhancement* regarding support for the OWL/XML format.

BTW, just successfully converted this file with http://mowl-power.cs.man.ac.uk:8080/converter/

**Comment by carueda** [comment]
Implemented in ORR Portal 2.5.8 (201511240101).

<img width="675" alt="2015-11-23_2305" src="https://cloud.githubusercontent.com/assets/556505/11360191/1a4bcdb2-9237-11e5-9651-71de12192833.png">

This uses the OWL API library to load the uploaded file and internally convert it to RDF/XML to then simply use Jena to load the model, and then continue the registration sequence as usual.

Note: Using OWL API v3.4.5 (2013-07-26); I tried a much more recent version but it brings 
*a lot* of dependencies and, worse, make the whole ORR application fail to load or 
operate normally possibly because of conflicts with other already used libraries.
Didn't investigate much but once I noted that http://mowl-power.cs.man.ac.uk:8080/converter/
worked fine to convert the ice-of-land-origin ontology and that it said it used OWL API 
Version 3.4.5-SNAPSHOT, I decided just to try the 3.4.5 version available from the sourceforge 
download page. Only adding owlapi-distribution-3.4.5.jar to the classpath was enough.
